processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java

```diff
@@ -196,7 +246,7 @@ static BaseFloatColumnValueSelector makeColumnValueSelectorWithFloatDefault(
     if (fieldName != null) {
       return metricFactory.makeColumnValueSelector(fieldName);
     } else {
-      final Expr expr = Parser.parse(fieldExpression, macroTable);
+      final Expr expr = parseIfAbsent(fieldExpression, macroTable);
```

Comment:
would be better to not have explicit cache added in this class. Instead, if you change arguments of this(and other similar methods) to...

```
  static BaseFloatColumnValueSelector makeColumnValueSelectorWithFloatDefault(
      final ColumnSelectorFactory metricFactory,
      @Nullable final String fieldName,
      @Nullable final Expr fieldExpression,
      final float nullValue
  )
```

and let caching happen in `SimpleXXAggregatorFactory` classes . Don't parse in the constructor of those classes because those objects could be created in many places where they wouldn't need parsed expression, so do the parsing lazily.

Reply:
@himanshug Good advice! I'll push a new commit when I resolve it.

Reply:
@himanshug Would it be better if we move the cache to the Parser class and add a new method like parseIfAbsent? Otherwise, we need to add cache in  SimpleXXAggregatorFactory, respectively.

```
public abstract class SimpleFloatAggregatorFactory extends NullableAggregatorFactory<BaseFloatColumnValueSelector>
{
  protected final String name;
  @Nullable
  protected final String fieldName;
  @Nullable
  protected final String expression;
  protected final ExprMacroTable macroTable;
  protected final Supplier<Expr> fieldExpression;

  public SimpleFloatAggregatorFactory(
      ExprMacroTable macroTable,
      String name,
      @Nullable final String fieldName,
      @Nullable String expression
  )
  {
    this.macroTable = macroTable;
    this.name = name;
    this.fieldName = fieldName;
    this.expression = expression;
    this.fieldExpression = () -> expression == null ? null : Parser.parseIfAbsent(expression, macroTable);
```

Reply:
sent a PR to your branch, pls merge that. I meant to remove cache altogether and tie lifetime of parsed expression to the object holding it.

Reply:
New commit is pushed. 
Remove the loading cache.

Reply:
@himanshug sorry for the delayed update! Can you review this patch again?

Reply:
@viongpanzi  changes look good, thanks.

also, not sure if you know/understood but when I raised https://github.com/viongpanzi/incubator-druid/pull/1 , I would expect you to merge the PR instead of copying the changes so that original commit stays alongside your commits . something for next time :)

Reply:
@himanshug Thanks for your review. Sorry for the misunderstood! I need to learn more about github and I'll pay attention next time.